Update docs for PipeReader.TryRead

waf opened the pull request:
As per https://github.com/dotnet/runtime/issues/30161#issuecomment-831035626, add some documentation to `PipeReader.TryRead` to better describe some unclear behavior.

The unclear behavior technically resides in the `StreamPipeReader` implementation, which is internal, so I added this as a remark to the public `PipeReader.TryRead` method instead, so it will show up in the public documentation.

waf committed:
Update docs for PipeReader.TryRead

As per https://github.com/dotnet/runtime/issues/30161#issuecomment-831035626, add some documentation to `PipeReader.TryRead` to better document some unclear behavior.

The unclear behavior technically resides in the `StreamPipeReader.TryRead` implementation, which is internal, so I added this as a `<remark>` to the public `PipeReader.TryRead` method instead.

omariom commented on src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs:
```diff
@@ -15,7 +15,9 @@ public abstract partial class PipeReader
-        /// <remarks>If the pipe returns <see langword="false" />, there is no need to call <see cref="System.IO.Pipelines.PipeReader.AdvanceTo(System.SequencePosition,System.SequencePosition)" />.</remarks>
+        /// <remarks>If the pipe returns <see langword="false" />, there is no need to call <see cref="System.IO.Pipelines.PipeReader.AdvanceTo(System.SequencePosition,System.SequencePosition)" />.
+        /// If the current <see cref="System.IO.Pipelines.PipeReader" /> implementation is a <see cref="System.IO.Pipelines.StreamPipeReader" />,
+        /// <see cref="System.IO.Pipelines.PipeReader.ReadAsync(System.Threading.CancellationToken)" /> must be called first in order to asynchronously retrieve data from the <see cref="System.IO.Pipelines.PipeReader" />.</remarks>
```
PipeReader is an abstraction that can have many implementations. The documentation should not mention just one of them.
Would be better to describe what it does (reads unconsumed bytes of the last ReadAsync?) 
and the proper usage which should work with any conforming implementation.

waf replied:
Thanks, I agree that it's somewhat awkward to mention a specific implementation here. I'll update this.

waf replied:
@omariom The more I think this through, the more I think that the current approach (or one like it) of being specific to StreamPipeReader is the correct approach. I'm happy to put this documentation in a different place, though, if this method doc is not the best place.

Here's my reasoning:

1. I believe the intention of `PipeReader.TryRead` is support reading synchronously (i.e. not in conjunction with `ReadAsync`). This is according to the current method `<summary>` documentation, as well as implementations [here](https://github.com/mgravell/Pipelines.Sockets.Unofficial/blob/19b141f30fe2485dd79993be0ac87a9fc6b34b74/src/Pipelines.Sockets.Unofficial/MemoryMappedPipeReader.cs#L217) and [here](https://github.com/dotnet/runtime/blob/main/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/SequencePipeReader.cs#L74).
2. I think StreamPipeReader does not have the same semantics, but it's a very popular implementation. See https://stackoverflow.com/questions/64820259 for some of the confusion I'm trying to clear up.

I don't think we should document StreamPipeReader's current semantics as the intended semantics of the base class. Any thoughts on the above? Thanks.

halter73 replied:
I was tempted to ask for this to be put on PipeReader.Create, but I think it's better here where it is because this is really a gotcha for user's PipeReader.TryRead who very well might not know where the PipeReader came from. ASP.NET Core's HttpContext.Request.BodyReader is sometimes backed by StreamPipeReader and sometimes backed by DefaultPipeReader for example.

That said, StreamPipeReader is an internal type, so we shouldn't reference it directly.

How about something like this instead with the appropriate `<see ...` markup?

> The PipeReader implementation returned by `PipeReader.Create(Stream, StreamPipeReaderOptions?)` will not read new data from the backing Stream when TryRead is called. PipeReader.ReadAsync must be called to read new data from the backing Stream. Any unconsumed data from a previous read should always be available from `TryRead`.

Also, can we make this a "note" or something so it stands apart from the rest of the docs for the method which are more about general usage?

waf replied:
Thanks @halter73, that's a good way to avoid mentioning the internal type. I've pushed a change that's basically your wording, with the final sentence tweaked a bit to clearly differentiate between the async/sync read calls. Thanks!

davidfowl commented:
It's possible that TryRead should throw or block on `StreamPipeReader` instead of this documentation quirk. Feels a bit better than letting your application go into an infinite loop.

waf committed:
remove doc reference to internal StreamPipeReader

The StreamPipeReader <see .. /> reference is an internal type, so it should not be mentioned in the public documentation.

halter73 commented on src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs:
```diff
@@ -15,7 +15,15 @@ public abstract partial class PipeReader
+        /// <remarks>
+        /// If the pipe returns <see langword="false" />, there is no need to call <see cref="System.IO.Pipelines.PipeReader.AdvanceTo(System.SequencePosition,System.SequencePosition)" />.
+        /// <para>
+        /// Note: the <see cref="System.IO.Pipelines.PipeReader" /> implementation returned by <see cref="System.IO.Pipelines.PipeReader.Create(System.IO.Stream, System.IO.Pipelines.StreamPipeReaderOptions?)"/>
```
@dotnet/docs do we need to quote this and add a `[!NOTE]` for this to show up correctly?

I think so looking at [this](https://github.com/dotnet/dotnet-api-docs/blob/main/xml/System.Threading/ThreadPool.xml#L85-L91)

adegeo replied:
I don't know that you can use `[!NOTE]` because that's markdown. /// comments are HTML. I believe the only way to use markdown and the ms.docs syntax extensions (which `[!NOTE]` is) is to migrate this content to the external XML API files.

Tagging @BillWagner @gewarren for comment.

gewarren replied:
You'll need to wrap it in `<format>` tags like this: https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeWriter.cs#L46.

jeffhandley committed:
Merge branch 'main' into update-docs-for-streampipereader-tryread

jeffhandley committed:
Format remarks as markdown and use !IMPORTANT

jeffhandley committed:
Merge branch 'main' into update-docs-for-streampipereader-tryread
